File: 2024/day22.cpp
#include <string>
#include <vector>
#include <unordered_map>

#include "../util.hpp"

inline int getNextSecret(int x){
    x = ((x ^ (x << 6)) & 0xffffff);
    x = ((x ^ (x >> 5)) & 0xffffff);
    x = ((x ^ (x << 11)) & 0xffffff);
    return x;
}

int main(int, char**){
    std::vector<int> initSecrets = util::getInts(util::getInput(2024, 22));
    unsigned long long int part1Sum = 0;
    std::unordered_map<int, int> totalCounter, counter;
    for (int s : initSecrets){
        counter.clear();
        int coor = 0, oldPrice = s % 10;
        for (int c = 1; c <= 2000; ++c){
            int ss = getNextSecret(s);
            int price = ss % 10;
            coor = coor / 19 + (price - oldPrice + 9) * 6859;  // 19 ** 3 == 6859
            if (c >= 4 && !util::contains(coor, counter)){
                counter[coor] = price;
            }
            oldPrice = price;
            s = ss;
        }
        part1Sum += s;
        for (auto [coor, val] : counter){
            totalCounter[coor] += val;
        }
    }
    util::output(part1Sum);
    int maxval = -1;
    for (auto [coor, val] : totalCounter){
        if (val > maxval){
            maxval = val;
        }
    }
    util::output(maxval);

    return 0;
}

File: 2024/day22.py
import AOCInit
import util

from collections import Counter

if __name__ != '__main__':
    exit()

inp = util.getInput(d=22, y=2024)

initSecrets = tuple(int(x) for x in inp.splitlines())

pruneMask = 16777216 - 1

def getNextSecret(x):
    x = (x ^ (x << 6)) & pruneMask
    x = (x ^ (x >> 5)) & pruneMask
    x = (x ^ (x << 11)) & pruneMask
    return x

part1Sum = 0
totalCounter: Counter[int] = Counter()
for s in initSecrets:
    counter: Counter[int] = Counter()
    oldPrice = s % 10
    coor = 0
    for i in range(2000):
        ss = getNextSecret(s)
        price = ss % 10
        coor = coor // 19 + (price - oldPrice + 9) * (19 ** 3)
        if i >= 3 and coor not in counter:
            counter[coor] = price
        s = ss
        oldPrice = price
    part1Sum += s
    totalCounter += counter
print(part1Sum)
print(totalCounter.most_common(1)[0][1])

